From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: EJ Hsu <ejh@nvidia.com>
Date: Mon, 12 Aug 2024 09:18:19 +0000
Subject: [PATCH 136/243] NVIDIA: SAUCE: xhci: tegra: fix memory corruption

In XUSB virtualization, the padctl interrupt will be directed to
xhci server. In order for VF to get notified when padctl interrupt
is triggered, VF needs to register a notifier to padctl driver during
its probe().

Add tegra_xusb_padctl_event_unregister() so that VF can remove its
notifier when unbind operation is called.

http://nvbugs/4762786

Change-Id: Icfff7cf42d4d4c72e4569ecbd974ec6b15b8b171
Signed-off-by: EJ Hsu <ejh@nvidia.com>
(cherry picked from commit 5d627a28b5fef0f0781b457ea898b9b1570a5737)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3201124
Tested-by: Petlozu Pravareshwar <petlozup@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Petlozu Pravareshwar <petlozup@nvidia.com>

diff --git a/drivers/usb/host/xhci-tegra.c b/drivers/usb/host/xhci-tegra.c
index 06a02e65886f..6b6bba97385b 100644
--- a/drivers/usb/host/xhci-tegra.c
+++ b/drivers/usb/host/xhci-tegra.c
@@ -2587,13 +2587,6 @@ static int tegra_xusb_probe(struct platform_device *pdev)
 		}
 	}
 
-#if defined(PHY_TEGRA_XUSB_SUPPORT_EVENT)
-	if (tegra->soc->is_xhci_vf) {
-		tegra->padctl_nb.notifier_call = tegra_xhci_padctl_notify;
-		tegra_xusb_padctl_event_register(tegra->padctl, &tegra->padctl_nb);
-	}
-#endif
-
 	err = tegra_xusb_enable_firmware_messages(tegra);
 	if (err < 0) {
 		dev_err(&pdev->dev, "failed to enable messages: %d\n", err);
@@ -2606,6 +2599,17 @@ static int tegra_xusb_probe(struct platform_device *pdev)
 		goto remove_usb3;
 	}
 
+#if defined(PHY_TEGRA_XUSB_SUPPORT_EVENT)
+	if (tegra->soc->is_xhci_vf) {
+		tegra->padctl_nb.notifier_call = tegra_xhci_padctl_notify;
+		err = tegra_xusb_padctl_event_register(tegra->padctl, &tegra->padctl_nb);
+		if (err < 0) {
+			dev_err(&pdev->dev, "failed to register padctl event: %d\n", err);
+			goto remove_usb3;
+		}
+	}
+#endif
+
 	/* Enable wake for both USB 2.0 and USB 3.0 roothubs */
 	device_init_wakeup(&tegra->hcd->self.root_hub->dev, true);
 	device_init_wakeup(&xhci->shared_hcd->self.root_hub->dev, true);
@@ -2670,6 +2674,11 @@ static void tegra_xusb_remove(struct platform_device *pdev)
 	struct tegra_xusb *tegra = platform_get_drvdata(pdev);
 	struct xhci_hcd *xhci = hcd_to_xhci(tegra->hcd);
 
+#if defined(PHY_TEGRA_XUSB_SUPPORT_EVENT)
+	if (tegra->soc->is_xhci_vf)
+		tegra_xusb_padctl_event_unregister(tegra->padctl, &tegra->padctl_nb);
+#endif
+
 	tegra_xusb_deinit_usb_phy(tegra);
 
 	pm_runtime_get_sync(&pdev->dev);
-- 
2.43.0

